core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -51,8 +185,744 @@ trait DataSkippingReader
   private lazy val withStatsCache =
     cacheDS(withStatsInternal0, s"Delta Table State with Stats #$version - $redactedPath")
 
-  // TODO(scott): proper implementation
+  protected def withStatsInternal: DataFrame = withStatsCache.getDS
+
+  /** All files with the statistics column dropped completely. */
+  def withNoStats: DataFrame = allFiles.drop("stats")
+
+  /**
+   * Returns a parsed and cached representation of files with statistics.
+   *
+   *
+   * @return cached [[DataFrame]]
+   */
+  final def withStats: DataFrame = {
+    withStatsInternal
+  }
+
+
+  // Helper method for expression types that represent an IN-list of literal values.
+  //
+  //
+  // For excessively long IN-lists, we just test whether the file's min/max range overlaps the range
+  // spanned by the list's smallest and largest elements.
+  private def constructLiteralInListDataFilters(a: Expression, possiblyNullValues: Seq[Any]):
+      Option[DataSkippingPredicate] = {
+    val dt = a.dataType
+    // The Ordering we use for sorting cannot handle null values, and these can anyway
+    // be safely ignored because they will never cause an IN-list predicate to return TRUE.
+    val values = possiblyNullValues.filter(_ != null)
+    lazy val ordering = TypeUtils.getInterpretedOrdering(dt)
+    if (values.isEmpty) {
+      // Handle the trivial empty case even for otherwise ineligible types.
+      // NOTE: SQL forbids empty in-list, but InSubqueryExec could have an empty subquery result
+      // or IN-list may contain only NULLs.
+      Some(DataSkippingPredicate(falseLiteral))
+    } else if (!SkippingEligibleLiteral.isEligibleDataType(dt)) {
+      // Don't waste time building expressions for incompatible types
+      None
+    }
+    else {
+      // Emit filters for an imprecise range test that covers the entire entire list.
+      val min = Literal(values.min(ordering), dt)
+      val max = Literal(values.max(ordering), dt)
+      constructDataFilters(And(GreaterThanOrEqual(max, a), LessThanOrEqual(min, a)))
+    }
+  }
+
+  /**
+   * Returns a file skipping predicate expression, derived from the user query, which uses column
```

Comment:
+1000... thank you for these detailed inline docs. Would have been much harder to review this large PR without this.